x/dex/keeper/limitorder.go

```diff
@@ -0,0 +1,231 @@
+	// memoize prices
+	var priceOutToIn sdk.Dec
+	var err error
+	if tokenIn == token0 {
+		priceOutToIn, err = CalcPrice1To0(tickIndex)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		priceOutToIn, err = CalcPrice0To1(tickIndex)
+		if err != nil {
+			return nil, err
+		}
+	}
+	priceInToOut := sdk.OneDec().Quo(priceOutToIn)
+
+	return &LimitOrderTranche{
+		types.LimitOrderTranche{
+			TrancheIndex:     trancheIndex,
+			TickIndex:        tickIndex,
+			TokenIn:          tokenIn,
+			PairId:           pairId,
+			ReservesTokenIn:  reservesIn,
+			ReservesTokenOut: reservesOut,
+			TotalTokenIn:     totalIn,
+			TotalTokenOut:    totalOut,
+		},
+		priceInToOut,
+		priceOutToIn,
+	}, nil
+}
+
+// Zero-initialized tranche object constructor
+func NewLimitOrderTrancheDefault(tokenIn string, token0 string, pairId string, tickIndex int64, trancheIndex uint64) (*LimitOrderTranche, error) {
+	return NewLimitOrderTranche(
+		tokenIn,
+		token0,
+		pairId,
+		tickIndex,
+		trancheIndex,
+		sdk.ZeroInt(),
+		sdk.ZeroInt(),
+		sdk.ZeroInt(),
+		sdk.ZeroInt(),
+	)
+}
+
+// Fetch tranche from keeper's KVStore and return error if not found
+func NewLimitOrderTrancheFromKeeper(ctx sdk.Context, k Keeper, tokenIn string, token0 string, pairId string, tickIndex int64, trancheIndex uint64) (tranche *LimitOrderTranche, err error) {
+	trancheProto, found := k.GetLimitOrderTranche(ctx, pairId, tickIndex, tokenIn, trancheIndex)
+	if !found {
+		return nil, types.ErrValidLimitOrderMapsNotFound
+	}
+	tranche, err = NewLimitOrderTranche(
+		trancheProto.TokenIn,
+		token0,
+		trancheProto.PairId,
+		trancheProto.TickIndex,
+		trancheProto.TrancheIndex,
+		trancheProto.ReservesTokenIn,
+		trancheProto.ReservesTokenOut,
+		trancheProto.TotalTokenIn,
+		trancheProto.TotalTokenOut,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return tranche, nil
+}
+
+// Fetch tranche from keeper's KVStore and zero-initialize if not found
+func NewInitLimitOrderTrancheFromKeeper(ctx sdk.Context, k Keeper, tokenIn string, token0 string, pairId string, tickIndex int64, trancheIndex uint64) (tranche *LimitOrderTranche, err error) {
+	tranche, err = NewLimitOrderTrancheFromKeeper(ctx, k, tokenIn, token0, pairId, tickIndex, trancheIndex)
+	// if not found in KVStore, call zero-initializing constructor
+	if err == types.ErrValidLimitOrderMapsNotFound {
+		tranche, err = NewLimitOrderTrancheDefault(tokenIn, token0, pairId, tickIndex, trancheIndex)
+		if err != nil {
+			return nil, err
+		}
+		k.SetLimitOrderTranche(ctx, tranche.LimitOrderTranche)
+	} else if err != nil {
+		return nil, err
+	}
+	return tranche, nil
+}
+
+func (tranche *LimitOrderTranche) Save(ctx sdk.Context, k Keeper) {
+	k.SetLimitOrderTranche(ctx, tranche.LimitOrderTranche)
+}
+
+func (tranche *LimitOrderTranche) PartiallyFilled() bool {
```

Comment:
PartiallyFilled or AtLeastPartiallyFilled? Maybe Wet() as an analogy for a container that is at least partially filled, but maybe fully filled.